### ISSUE TITLE
Use events to couple parameters in instrument drivers?

### ISSUE THREAD

**Issue by alexcjohnson** [microsoft/Qcodes#92]
@MerlinSmiles noticed an interesting case while writing an instrument driver: sometimes different parameters are [coupled to each other](https://github.com/qdev-dk/Qcodes/pull/74#discussion_r60036178), such that changing one affects the value of the other(s). This means that the snapshot of this instrument will _definitely_ be incorrect right after a `set` call on one of these parameters.

Sometimes this is just two different ways to call the same hardware command, eg resolution and averaging time... sometimes you want to specify one, sometimes the other, but they both generate identical hardware calls. In that case, we could probably work out a way to determine the value of the other parameter programmatically, but it would be easiest to just ask the instrument.

Other times they are separate parameters that don't have a 1:1 mapping but they impose constraints on each other. In that case, unless we want to figure out the details of these constraints and hard code them into the driver (which sounds horrible!) you _need_ to ask the instrument for the other parameter(s).

It occurred to me the easiest way to build in such couplings would be through events. In this case we would do something like:

``` python
def update_res_and_time(self):
    self.resolution.get()
    self.integration_time.get()
    self.NPLC.get()

def __init__(self, ...):
    self.add_parameter('resolution', ...)
    self.add_parameter('integration_time', ...)
    self.add_parameter('NPLC', ...)

    self.on(('resolution_set', 'integration_time_set', 'NPLC_set'), self.update_res_and_time)
```

Where the parameter would know, because it's attached to this instrument, to emit a `'<name>_set'` event on setting (and maybe a `'<name>_get'` event on getting too?) on this instrument's event emitter.

But I haven't used any event systems in Python... anyone? I see [pyee](https://pypi.python.org/pypi/pyee), [gevent-eventemitter](https://pypi.python.org/pypi/gevent-eventemitter), [eventemitter](https://pypi.python.org/pypi/eventemitter/0.2.0)...

Is this the right way to do this or is there something better? Do we need such events to extend beyond the walls of the instrument (which in practice would also mean extending beyond this PROCESS, which would make it quite a bit more involved)? pyee looks nice and lightweight, people seem to like it, so that's what I would go with all else equal.

**Comment by giulioungaretti** [comment]
@spauka @MerlinSmiles    not sure I can follow this conversation, Alex link is broken.
Could you summarise // point out if it's still relevant   ?

**Comment by spauka** [comment]
The issue that is being referred to is still outstanding, in that as far as I can tell there is still no way of creating coupled parameters, though the solution referred to in the title, "Use events ...", may no longer be the best solution.

In short, the issue is that some instruments expose certain parameters only sometimes, or that the values/units of other parameters change depending on the selection of a parameter. For example, looking at the [SR830 driver](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/stanford_research/SR830.py#L294), when we change the value of [`input_config`](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/stanford_research/SR830.py#L116), the units of `X`,`Y` and `sensitivity` can change from voltage(V) to current(I). The way that this is currently handled is that the get/set parser of `input_config` is overriden to change the value of the units, however arguably this is not the role of the get/set parser function. Other times, certain parameters only take a value if another parameter is enabled, see my previous comment re. averages. I would say these are the two issues noted in Alex's link.

Don't have a strong opinion on the best solution, but moving to a more logical way of handling these parameters should be addressed.

**Comment by giulioungaretti** [comment]
@spauka  yes agree!  @WilliamHPNielsen had a lot of the same issues writing drivers (correct?)

**Comment by WilliamHPNielsen** [comment]
@giulioungaretti Yes, that is correct, we had that a lot with the ZI UHF-LI driver.  If you want to convince yourself that Alex and Sebastian are right that hardcoding this stuff into `set` and `get` is not the proper way to go about it, have a look at this highly specialised `set` function: https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/ZI/ZIUHFLI.py#L1467. 

I fully agree that we should solve this. Although the SCPI standard encourages tree-like disparate parameters, most instruments still have them mingled together.

**Comment by giulioungaretti** [comment]
@WilliamHPNielsen @spauka  yeah, suspected so just wanted to make sure.

**Comment by nulinspiratie** [comment]
Solving coupled parameters with events sounds like a good idea to me. We use the package [blinker](https://pypi.python.org/pypi/blinker) for events in our qcodes extension, and it works quite nicely. It's quite fast, and allows both named and anonymous signals. A big advantage of blinker is that it allows to subscribe to specific senders.

As for implementation, one possibility is to attach an event object to each parameter. Simple get/set signals can then be hardcoded into the get/set parameters, for instance in decorators suggested in #600. @WilliamHPNielsen signals can also be used for more complicated cases such as the one you mentioned, by sending signals that don't have get or set as a kwarg, but other specific ones. Would this be a viable solution?

**Comment by giulioungaretti** [comment]
@nulinspiratie  does it work across processes?

**Comment by nulinspiratie** [comment]
@giulioungaretti No it doesn't, I'm not aware of any event handler that also transfers signals across processes. A wrapper could probably be programmed around the package that could handle this, but I'm not sure how much overhead this would give
